lib/caption-stream.js

```diff
@@ -0,0 +1,248 @@
+    if (String.fromCharCode(sei.payload[3],
+                            sei.payload[4],
+                            sei.payload[5],
+                            sei.payload[6]) !== 'GA94') {
+      return null;
+    }
+
+    // finally, user_data_type_code should be 0x03 for caption data
+    if (sei.payload[7] !== 0x03) {
+      return null;
+    }
+
+    // return the user_data_type_structure and strip the trailing
+    // marker bits
+    return sei.payload.subarray(8, sei.payload.length - 1);
+  };
+
+  // see CEA-708-D, section 4.4
+  var parseCaptionPackets = function(pts, userData) {
+    var results = [], i, count, offset, data;
+
+    // if this is just filler, return immediately
+    if (!(userData[0] & 0x40)) {
+      return results;
+    }
+
+    // parse out the cc_data_1 and cc_data_2 fields
+    count = userData[0] & 0x1f;
+    for (i = 0; i < count; i++) {
+      offset = i * 3;
+      data = {
+        type: userData[offset + 2] & 0x03,
+        pts: pts
+      };
+
+      // capture cc data when cc_valid is 1
+      if (userData[offset + 2] & 0x04) {
+        data.ccData = (userData[offset + 3] << 8) | userData[offset + 4];
+        results.push(data);
+      }
+    }
+    return results;
+  };
+
+  var CaptionStream = function() {
+    CaptionStream.prototype.init.call(this);
+
+    this.field1_ = new Cea608Stream();
+    this.field1_.on('data', this.trigger.bind(this, 'data'));
+  };
+  CaptionStream.prototype = new muxjs.Stream();
+  CaptionStream.prototype.push = function(event) {
+    var sei, userData, captionPackets, i;
+
+    // only examine SEI NALs
+    if (event.nalUnitType !== 'sei_rbsp') {
+      return;
+    }
+    // parse the sei
+    sei = parseSei(event.data);
+
+    // ignore everything but user_data_registered_itu_t_t35
+    if (sei.payloadType !== 4) {
+      return;
+    }
+
+    // parse out the user data payload
+    userData = parseUserData(sei);
+
+    // ignore unrecognized userData
+    if (!userData) {
+      return;
+    }
+
+    // parse out CC data packets
+    captionPackets = parseCaptionPackets(event.pts, userData);
+
+    // send the data to the appropriate field
+    for (i = 0; i < captionPackets.length; i++) {
+      if (captionPackets[i].type === 0) {
+        this.field1_.push(captionPackets[i]);
+      }
+    }
+  };
+
+  // ----------------------
+  // Session to Application
+  // ----------------------
+
+  var BASIC_CHARACTER_TRANSLATION = {
+    0x5c: 0xe9,
+    0x5e: 0xed,
+    0x5f: 0xf3,
+    0x60: 0xfa,
+    0x7b: 0xe7,
+    0x7c: 0xf7,
+    0x7d: 0xd1,
+    0x7e: 0xf1,
+    0x2a: 0xe1,
+    0x7f: 0x2588
+  };
+
+  var Cea608Stream = function() {
+    Cea608Stream.prototype.init.call(this);
+
+    this.startPts_ = 0;
+    this.displayed_ = '';
+    this.nonDisplayed_ = '';
+
+    this.push = function(packet) {
+      var data, swap, charCode;
+      // remove the parity bits
+      data = packet.ccData & 0x7f7f;
+
+      switch (data) {
+      case 0x0000:
```

Comment:
My feeling is pretty much that you shouldn't need comments. But that is a stylistic preference. I would have a big object named `608PacketType` or somesuch with a nice name -> number mapping for the packets because I feel it makes it simpler to understand what it going on (considering it is already a not-straight forward state-machine type thing.)